SMS support for received & read & failed stats

kuzmany opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:
This pull request introduces support for tracking delivered, read, and failed statistics in SMS messages. By default, it should function as it did before. However, it adds new columns to the sms_message and sms_messages tables. 

Any SMS transport integration can enhance the transport by incorporating additional interfaces to enable these statistics in the user interface. The transport itself should be responsible for managing the incrementing of each column's count. In the future, we may consider adding some services to simplify this process.

In another PR we can bring some helpers to easier manage data for transport.


<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2.  Run migration 
3.  Check If everything run as before  without changes
4.  Then dev during the code review can enabled Twilio with fake authorization data an add to Twilio transport new interfaces:

`class TwilioTransport implements TransportInterface, DeliverySupportInterface, ReadSupportInterface, FailedSupportInterface`


![image](https://github.com/mautic/mautic/assets/462477/2206f977-6dc4-456f-91c2-93fe947fdd32)

5. Then create SMS message
6. See on list page the new stats column
7. See on detail page new line graph options

![image](https://github.com/mautic/mautic/assets/462477/653c7ed8-e05e-438e-b6d9-aa5abe3a20b1)

![image](https://github.com/mautic/mautic/assets/462477/34e06450-1bb7-472a-b498-553eae35f2d2)



<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

kuzmany committed:
SMS support callback with received & read & failed notifications

kuzmany committed:
Remove COLUMN constant from DeliverySupportInterface, ReadSupportInterface, and FailedSupportInterface.

kuzmany committed:
Update SMS list_stats_column.html.twig and en_US/messages.ini files to display the read count and ratio of text messages, add new translations for delivered and failed counts.

kuzmany committed:
Refactor buildSmsQuery, buildSmsSentQuery, buildSmsDeliveredQuery, buildSmsReadQuery, and buildSmsFailedQuery to include return type declaration.

kuzmany committed:
Refactor getDeliveredRatio(), getReadRatio(), and getFailedRatio() in Sms.php to return float instead of int.

kuzmany committed:
Resolve CS issue

kuzmany committed:
Add AvailableStatsDTOTest.php to the SmsBundle/Tests/Stat/DTO directory.

kuzmany committed:
Fix code styles check

kuzmany committed:
Merge branch '5.x' into sms-callback-support-m5

kuzmany committed:
Add functional test for the SMS controller's list view with assertions.

kuzmany committed:
Refactor SMS list stats column links to use correct search command.

kuzmany committed:
Add setupTwilio() method to SmsControllerFunctionalTest.

kuzmany committed:
Merge branch '5.x' into sms-callback-support-m5

kuzmany committed:
Remove extra 'g' in assertStringNotContainsString function calls.

kuzmany committed:
Update SMS ID references in assertStringContainsString and assertStringNotContainsString functions.

mabumusa1 commented:
I did a code review the changes are sound, I will test it over the weekend with campaign to see the results

kuzmany committed:
Merge branch '5.x' into sms-callback-support-m5

RCheesley committed:
Merge branch '5.x' into sms-callback-support-m5

RCheesley commented:
Anyone wanting to test this, I just made a PR to add the Twilio configuration docs: https://github.com/mautic/user-documentation/pull/260

kuzmany committed:
Merge branch 'refs/heads/5.x' into sms-callback-support-m5

# Conflicts:
#	app/bundles/SmsBundle/EventListener/LeadSubscriber.php

kuzmany committed:
CS fix

RCheesley committed:
Merge branch '5.x' into sms-callback-support-m5

RCheesley commented:
Blocked from testing this due to https://github.com/mautic/mautic/issues/13646 - assume that nobody can send SMS messages in campaigns in Mautic 5 if this is happening.

kuzmany committed:
Merge remote-tracking branch 'upstream/7.x' into sms-callback-support-m5

# Conflicts:
#	app/bundles/SmsBundle/Controller/AjaxController.php
#	app/bundles/SmsBundle/Resources/views/Sms/list_stats.html.twig

kuzmany committed:
style: Update SMS stats column labels and styling

kuzmany committed:
test: Remove unnecessary parentheses in anonymous class definition

RCheesley committed:
Merge branch '7.x' into sms-callback-support-m5

kuzmany committed:
Merge branch '7.x' into sms-callback-support-m5

kuzmany committed:
Fix duplicate class declaration error

Remove duplicate SMSControllerFunctionalTest.php file that was causing
PHP Fatal error due to redeclared class name. Keep the newer
SmsControllerFunctionalTest.php with proper naming convention.

kuzmany committed:
Revert "Fix duplicate class declaration error"

This reverts commit 2f73a6db6744b8b4981d801757b22fcc436ee009.

kuzmany committed:
Rename test file to correct case and remove duplicate

- Move SmsControllerFunctionalTest.php to SMSControllerFunctionalTest.php
- Remove the original lowercase filename to avoid naming conflicts

kuzmany committed:
Fix PHP CS Fixer formatting issue

- Add trailing newline to SMSControllerFunctionalTest.php

kuzmany committed:
Update SMS controller functional test with project support

- Add setUp method to configure site_url parameter
- Add testSmsWithProject test method for project association
- Update CreateSms method with default parameters and template type
- Add Project entity usage for testing SMS-project relationships

kuzmany committed:
Make SMS controller test class final and fix code style

- Add final keyword to SMSControllerFunctionalTest class
- Fix string concatenation spacing to match code style standards

kuzmany committed:
Add missing Project entity import
